### PR TITLE
LSP: Enforce only one popup at a time.

### PR DESCRIPTION
**Pull request by norcalli**
At first I was fixing the hover popups ( #11395 #11508 ):
- Close other popups on creation.
- Close popup on CursorMovedI.

Then I decided I might as well try this new method which I believe can be shared
- Add buffer_manager and use it for popups.
- Use buffer_manager for diagnostics.

Also I'm hiding the peekDefinition stuff since it's not quite ready for prime time. I'm thinking of changing it to read the file from disk if the buffer isn't available. It also seems like it's not possible with either the api.txt or eval.txt methods to specify the viewport of a floating buffer. Specifically, I want to be able to specify the rectange of how to view a buffer based on its cursor positions.

Also, where should we be putting stuff `buffer_manager` for the future? I have it under `vim.util` right now, but that's obviously a bit generic. I think @bfredl mentioned something from treesitter that could also benefit from it and I know I use that pattern a lot in colorizer/lsp. If you guys like it, I'll do more replacements with it.

Here's some thoughts I had from the gitter:
```
Trying to figure out where to put it. We still haven't added too much stuff to the stdlib. I'm tempted to start a `vim.util` but I don't know if that's too generic
Future things we're going to add are things like path, but that's easy since `vim.path` makes sense
or would it be `vim.util.path`
Then there's luafun related stuff, so would that be `vim.util.iter`?
Also future data structures like a Trie
`vim.util.trie` or `vim.trie`
```

### CODE REVIEW

**Comment by clason** [comment]
Regarding `peekDefinition`: That should work out of the box? vim-lsp does that based on buffers with neovim floating windows just fine (see the PR https://github.com/prabirshrestha/vim-lsp/pull/498).

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -306,10 +300,31 @@ function M.jump_to_location(location)
...
+}
+
+-- Close all popups which belong to us.
+function M.close_popups()
+  for _, v in popup_manager.iter() do
+    pcall(api.nvim_win_close, v.winnr, true)
+  end
+end
+
+-- Close all popups which belong to us.
```
Comment (norcalli) [inline runtime/lua/vim/lsp/util.lua]:
```suggestion
-- Close a popup by its name.
```

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -568,74 +639,65 @@ do
 
   function M.buf_clear_diagnostics(bufnr)
     validate { bufnr = {bufnr, 'n', true} }
-    bufnr = bufnr == 0 and api.nvim_get_current_buf() or bufnr
-    api.nvim_buf_clear_namespace(bufnr, diagnostic_ns, 0, -1)
+    -- bufnr = bufnr == 0 and api.nvim_get_current_buf() or bufnr
```
Comment (norcalli) [inline runtime/lua/vim/lsp/util.lua]:
```suggestion
```

**Comment by justinmk** [comment]
> Also, where should we be putting stuff `buffer_manager` for the future? I have it under `vim.util` right now, but that's obviously a bit generic.

No need for `vim.util`, that's the same thing as `vim.` itself.

> Future things we're going to add are things like path, but that's easy since `vim.path` makes sense or would it be `vim.util.path` ... `vim.util.iter` ... `vim.util.trie` 

Proposal: if a "family" of functions counts 10+, then it deserves a module. Otherwise it's just taxonomy fetish. So `vim.lsp` is justified. Maybe `vim.path` is justified, but `vim.url` probably isn't. 

We should err towards "too few" modules rather than too many. 

- As a user, it is easier to search one big module than many random categories chosen by taxonomy fetishists.
- As a developer, we can spend less time discussing what goes in which module.
- One can argue that "underscore prefixes" are a "poor man's module/namespace", but then when importing functions people end up adding prefixes anyways (e.g. to disambiguate `path_join` vs `url_join`).

File: runtime/lua/vim/util/buffer_manager.lua [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+--   on_detach(bufnr, lookup_value)
+--   on_lines
+--   on_changedtick
+--   utf_sizes
+-- }
+-- Returns: { attach(bufnr, ...), iter(), get(bufnr) }
+--
+-- Any arguments passed after bufnr to attach() are forwarded to buf_init()
+-- on_lines, on_changedtick, utf_sizes are forwarded to nvim_buf_attach
+local function buffer_manager(config)
```
Comment (justinmk) [inline runtime/lua/vim/util/buffer_manager.lua]: First line of the docstring should say qualitatively what it actually does / what is the purpose.

**Comment by tjdevries** [comment]
Hey @norcalli is this worth reviewing? I'll take a look at it if it would still be useful (just not sure if it is anymore or not since LSP moves so fast haha)

**Comment by tjdevries** [comment]
Hey @norcalli is this worth reviewing? I'll take a look at it if it would still be useful (just not sure if it is anymore or not since LSP moves so fast haha)

**Pull request closed without merging**